Objects/abstract.c

```diff
@@ -2426,31 +2426,51 @@ PyMapping_SetItemString(PyObject *o, const char *key, PyObject *value)
 }
 
 int
-PyMapping_HasKeyString(PyObject *o, const char *key)
-{
-    PyObject *v;
-
-    v = PyMapping_GetItemString(o, key);
-    if (v) {
-        Py_DECREF(v);
-        return 1;
+PyMapping_HasKeyString(PyObject *obj, const char *key)
+{
+    PyObject *dummy;
+    int rc = PyMapping_GetOptionalItemString(obj, key, &dummy);
+    if (rc < 0) {
+        _PyErr_WriteUnraisableMsg(
+            "in PyMapping_HasKeyString(); consider using "
+            "PyMapping_GetOptionalItemString() or PyMapping_GetItemString()",
+            NULL);
+        return 0;
     }
-    PyErr_Clear();
-    return 0;
+    // PyMapping_HasKeyString() also clears the error set before it's call
+    // if the key is not found.
+    if (rc == 0 && PyErr_Occurred()) {
+        _PyErr_WriteUnraisableMsg(
+            "in PyMapping_HasKeyString(); consider using "
+            "PyMapping_GetOptionalItemString() or PyMapping_GetItemString()",
```

Comment:
This case is very subtle and the error message is not heplful. It's non-obvious to me that the exception was already set *before* the function was called.

Maybe the error message should be something like: "Ignore exception set before calling in PyMapping_HasKeyString()".

Instead of "Exception ignored in PyMapping_HasKeyString()".

Reply:
Unfortunately `_PyErr_WriteUnraisableMsg()` does not support this. :frowning_face: